### DISCUSSION TITLE
refresh vs hard-refresh

### DISCUSSION THREAD

**Discussion by davidshen84** [argoproj/argo-cd#12237] [General]
Hi,

I understand the difference between sync vs refresh, but I could not figure out what's the difference between refresh vs hard-refresh.

I found that every time I updated my code, I have wait for the default 3 min reconsiliation time or hit the **hard-refresh** button. The **refresh** button or the `--refresh` option on CLI does not seem to do anything for me.

**Comment by jannfis** [comment]
* A normal refresh checks if your source has changed, e.g. if there was a new commit to Git, or if there's a new version of the Helm chart in the registry. This involves contacting the Git repository to ask for the latest commit ID of your tracking branch, or asking the Helm registry for the latest version of a Chart. If, and only if, the source has changed (e.g. a new commit or a new Helm chart version), the manifests will be re-rendered. A normal refresh happens every 3 minutes by default, and is requested by the application controller. 
* A hard refresh invalidates the manifest cache for your application, forcing the manifests to be re-rendered no matter whether there was a change in the source or not. One use case is e.g. when you have a Helm chart and change its code, without bumping its version number. A hard refresh usually occurs every 24 hours by default, when the manifest cache expires.

HTH.

**Reply by davidshen84** [reply]
So, normally, if I have a new commit, I should use the normal *refresh*. But somehow, I needed to use *hard-refresh* every time...

My source is on GitHub, so after I push the commits, the *refresh* function should get the latest commits immediately, right? Is there a log in the system that I can check if it got the latest commits when I trigger the *refresh* function?

**Reply by anthonyalayo** [reply]
I'm in the same boat as @davidshen84 and that is how I found this discussion. I push new git commits, but refresh never works for me, so I'm always using hard refresh.

**Reply by naviat** [reply]
I got the same also
@jannfis do we have any cases/solutions related to this issue?

**Reply by sankosa** [reply]
Hi @jannfis,

You wrote: "A hard refresh usually occurs every 24 hours by default, when the manifest cache expires"
Does it mean that we can change this hard refresh interval from 24 hours?
If yes, can you share with me the link how I can configure this?

Thank you in advance!

**Reply by amirsalaar** [reply]
I find that the refresh doesnt happen when I publish new helm charts (i.e, the helm chart digest changes in my registry -- the version of the chart doesn't change tho).

So, to get deployment start, I'd have to hit HARD refresh every time.

I am using argocd v2.10. Is there anyway that i can get this fixed?

**Reply by rumstead** [reply]
> I find that the refresh doesnt happen when I publish new helm charts (i.e, the helm chart digest changes in my registry -- the version of the chart doesn't change tho).
> 
> So, to get deployment start, I'd have to hit HARD refresh every time.
> 
> I am using argocd v2.10. Is there anyway that i can get this fixed?

That is intended. You need a change to your Argo CD application by updating the helm chart version to the new digest you just published. As @jannfis mentioned, the cache will invalidate every 24h by default and pick up any implicit changes you have.

**Reply by amirsalaar** [reply]
@rumstead but the actual helm chart version is not getting changed with new code push. you you mean just updating the helm values (for example image with the newly created image?)

**Reply by rumstead** [reply]
Are you tagging the new digest with your existing tag?

**Reply by jannfis** [reply]
Since a while, you can trigger a periodic hard refresh as well. It's a command line option to the argocd-application-controller pod:

```
--app-hard-resync int                    Time period in seconds for application hard resync.
```

You need to be careful though, because a hard refresh is a pretty expensive operation.

**Reply by mateusz-kleinert** [reply]
I would like to ask about the `--app-hard-resync` feature. @jannfis can you elaborate about it a little bit more? When I enabled it in my configuration set to 300 seconds I was able to spot some logs containing `requesting hard refresh`. However, it seems like the refresh was not triggered for all of my applications, just some more or less random subset. When I set it's value to 600 seconds I wasn't able to spot any logs suggesting that a hard refresh happened. Are there any other time constraints that may cause the hard refresh to not happen? I can't spot any logs that would suggest there is an error or some kind of a slow operation running in the background looking at the metrics.

**Comment by davidshen84** [comment]
Please try update your Argo images to the latest version. Also try to
increase resource requests. I found the refresh button works very well
recently.

**Comment by hirenko-v** [comment]
is it possible to configure per app?

**Comment by thiagowfx** [comment]
Another use case for hard refresh: when using the [helm-secrets](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration) <-> vault integration.

Updating a secret in Vault does not immediately reflect it in ArgoCD. A normal refresh doesn't propagate it either. Only a hard refresh propagates the updated Vault secret to ArgoCD.

Related:

- https://github.com/argoproj/argo-cd/issues/4002
- https://github.com/argoproj/argo-cd/issues/18000

**Comment by davidshen84** [comment]
For this particular case, the right way is using a version on your vault
key and updating the version when you update the key value.

**Comment by marinoborges** [comment]
I'm facing issues with Kustomize not refreshing after source change. It only shows the diff after a Hard Refresh. Is that a known issue?

**Reply by bensonphillipsiv** [reply]
Did you find a resolution to this?

**Reply by bensonphillipsiv** [reply]
I was using the `manifest-generate-paths` [annotation](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#manifest-paths-annotation) on my apps and removing it solved this issue.

**Reply by awm-achi** [reply]
We are also having this issue, we require path annotations since our repo is large. Is there a fix for this?
